instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java

```diff
@@ -0,0 +1,127 @@
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanBuilder;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.tracer.ClientSpan;
+import io.opentelemetry.instrumentation.api.tracer.ServerSpan;
+import java.util.List;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+// TODO(anuraaga): Need to define what are actually useful knobs, perhaps even providing a
+// base-class
+// for instrumentation library builders.
+/**
+ * An instrumenter of the start and end of a request/response lifecycle. Almost all instrumentation
+ * of libraries falls into modeling start and end, generating observability signals from these such
+ * as a tracing {@link Span}, or metrics such as the duration taken, active requests, etc. When
+ * instrumenting a library, there will generally be three steps.
+ *
+ * <ul>
+ *   <li>Create an {@link Instrumenter} using {@link InstrumenterBuilder}. Use the builder to
+ *       configure any library-specific customizations, and also expose useful knobs to your user.
+ *   <li>Call {@link Instrumenter#start(Context, Object)} at the beginning of a request.
+ *   <li>Call {@link Instrumenter#end(Context, Object, Object, Throwable)} at the end of a request.
+ * </ul>
+ */
+public class Instrumenter<REQUEST, RESPONSE> {
+
+  /** Returns a new {@link InstrumenterBuilder}. */
+  public static <REQUEST, RESPONSE> InstrumenterBuilder<REQUEST, RESPONSE> newBuilder(
+      OpenTelemetry openTelemetry,
+      String instrumentationName,
+      SpanNameExtractor<? super REQUEST> spanNameExtractor) {
+    return new InstrumenterBuilder<>(openTelemetry, instrumentationName, spanNameExtractor);
+  }
+
+  private final Tracer tracer;
+  private final SpanNameExtractor<? super REQUEST> spanNameExtractor;
+  private final SpanKindExtractor<? super REQUEST> spanKindExtractor;
+  private final StatusExtractor<? super REQUEST, ? super RESPONSE> statusExtractor;
+  private final List<? extends AttributesExtractor<? super REQUEST, ? super RESPONSE>> extractors;
+  private final ErrorCauseExtractor errorCauseExtractor;
+
+  Instrumenter(
+      Tracer tracer,
+      SpanNameExtractor<? super REQUEST> spanNameExtractor,
+      SpanKindExtractor<? super REQUEST> spanKindExtractor,
+      StatusExtractor<? super REQUEST, ? super RESPONSE> statusExtractor,
+      List<? extends AttributesExtractor<? super REQUEST, ? super RESPONSE>> extractors,
+      ErrorCauseExtractor errorCauseExtractor) {
+    this.tracer = tracer;
+    this.spanNameExtractor = spanNameExtractor;
+    this.spanKindExtractor = spanKindExtractor;
+    this.statusExtractor = statusExtractor;
+    this.extractors = extractors;
+    this.errorCauseExtractor = errorCauseExtractor;
+  }
```

Comment:
As long as this is package protected passing all the arguments is fine, but if we plan to make this public would suggest to have an `Option` pojo to allow us to add more arguments.

Reply:
This wouldn't become public, that's why we have the builder